src/Database/LSMTree/Internal/Run.hs

```diff
@@ -192,15 +183,16 @@ fromMutable :: HasFS IO h
             -> RunDataCaching
             -> RefCount
             -> RunBuilder RealWorld (FS.Handle h)
-            -> IO (Run RealWorld (FS.Handle h))
+            -> IO (Run IO (FS.Handle h))
 fromMutable fs hbio runRunDataCaching refCount builder = do
     (runRunFsPaths, runFilter, runIndex, runNumEntries) <-
       Builder.unsafeFinalise fs hbio (runRunDataCaching == NoCacheRunData) builder
-    runRefCount <- newMutVar refCount
     runKOpsFile <- FS.hOpen fs (runKOpsPath runRunFsPaths) FS.ReadMode
     runBlobFile <- FS.hOpen fs (runBlobPath runRunFsPaths) FS.ReadMode
     setRunDataCaching hbio runKOpsFile runRunDataCaching
-    return Run {..}
+    rec runRefCounter <- RC.unsafeMkRefCounterN refCount (Just $ close fs hbio r)
+        let !r = Run{..}
```

Comment:
Does this definitely work, being strict? I could imagine that we would need to avoid the `unsafeMkRefCounterN` trying to force the finaliser to WHNF.

Reply:
So, it appears that `rec` compiles to https://hackage.haskell.org/package/ghc-internal-9.1001.0/docs/src/GHC.Internal.System.IO.html#fixIO, which uses `MVar` for time travel -- and the core looks like it evaluates all the fields of the return `Run`.

Reply:
RE the `MVar` time travel, it roughly does the following
* Create MVar for the `Run`
* Create `RefCount` with a finaliser that reads the run from the `MVar`, and closes the run
* Create the run
* Put the run into the MVar